### DISCUSSION TITLE
Latest X64 version not able to setup in Windows 11 Pro ARM64

### DISCUSSION THREAD

**Discussion by Corsair-63** [Stellarium/stellarium#3023] [Q&A]
I've installed version Stellarium 0.22.1 x64 on Surface Pro X (Win 11 ARM64)
I've tried to update to latest version 1.22.3 x64 on the same tablet and always tell me that needs to be installed in a x64 system, as far as I know Win11 ARM64 has a x64 emulator and I'm running some x64 programs without any issue but it seems this software doeesn't properly recognise the emulation.

only thing please don't tell me to install x86 version, I won't walk backward ;-)

**Comment by 10110111** [comment]
> only thing please don't tell me to install x86 version, I won't walk backward ;-)

x86 and x86_64 are completely orthogonal to ARM and ARM64. If you're emulating, you're just going sideways, jumping from x86_64 to x86 is not backward.

> always tell me that needs to be installed in a x64 system

Is it the installer that says so? Could you show a screenshot?

> as far as I know Win11 ARM64 has a x64 emulator and I'm running some x64 programs without any issue

How exactly do you run them? Do you start the installer and the installed app the same way?

**Reply by Corsair-63** [reply]
<img width="546" alt="sshot-3" src="https://user-images.githubusercontent.com/76228638/214657152-00de016e-1e97-408a-8894-6cfaef91ceba.png">
hopefully I'll correctly attached screen-shot of installer after asking me the language of installation.

```
as far as I know Win11 ARM64 has a x64 emulator and I'm running some x64 programs without any issue

How exactly do you run them? Do you start the installer and the installed app the same way?``
```
when installing either by app either installer but after being installed by their menu shortcut without any problem like a normal app/software

**Comment by Corsair-63** [comment]
Also related in Issue:
https://github.com/Stellarium/stellarium/issues/2962

**Reply by 10110111** [reply]
So looks like you should try 1.2-qt5 instead of 1.2-qt6.

**Reply by Corsair-63** [reply]
<img width="813" alt="sshot-1" src="https://user-images.githubusercontent.com/76228638/214666249-af0bd794-3fae-4d73-8f10-9b1589c16999.png">
<img width="780" alt="sshot-2" src="https://user-images.githubusercontent.com/76228638/214666259-c99aed4b-ff89-4977-930d-3bb2798faf89.png">

Thanks a lot, working in x64 using 1.2-qt5

**Reply by gzotti** [reply]
Strange that Qt6 behaves worse than Qt5 here? However, I have no insight in Win11/ARM behaviour.

**Reply by Corsair-63** [reply]
It isn't different behaviour, it just x64 & qt6 version doesn't install on win11 ARM64 and x64 & qt5, yes

**Reply by 10110111** [reply]
Refusing to run and not refusing to run _is_ different behavior.

**Reply by 10110111** [reply]
@gzotti Does the installer itself differ between Qt5 and Qt6 builds? The error message shown seems to be generated by the Windows shell rather than the installer code. So the installer's EXE appears to be somehow different: maybe some PE header entries differ, or whatever.

**Reply by gzotti** [reply]
@alex-w does the packing. I think I remember having read that two different versions of the install program are used for Qt5- and Qt6-based builds. Maybe there are extra options to be activated?

**Reply by Corsair-63** [reply]
I've seen similar install behaviour with BCUnistaller and Win11 ARM64 not able to install it, but in the portable version you were able to "install" in HD and run x64 version too, I contacted with project administrator and he did an exchange in the installer and now "normal" version is properly installed in ARM64 systems.
I'm not a programmer and more important I could not give "expertise" advice to any one of you that are the real experts, you really know what could be done or suspect what could happens

**Reply by alex-w** [reply]
InnoSetup5 is using for Qt5-based packages and InnoSetup6 is using for Qt6-based packages. AppVeyor has arm64 version of Qt6 now and I can try building the new packages.

**Reply by alex-w** [reply]
> InnoSetup5 is using for Qt5-based packages and InnoSetup6 is using for Qt6-based packages. AppVeyor has arm64 version of Qt6 now and I can try building the new packages.

Qt6 for arm64 is not usable yet on Windows.

Please use Qt5-based packages for install Stellarium onto arm64 devices.

**Comment by hmartinez82** [comment]
I want to try to help building Stellarium for Windows on ARM64. I was about to ask where were the ci scripts for building the Windows x64 build are, but I found them. The AppVeyor one, correct? I have a few questions:
- Would it be possible to use MSYS2 as the build environment for compiling Stellarium for Windows on ARM64? There's a plethora of big applications being built for Windows on ARM64 using MSYS2 like Qt 5 and 6, Gimp, Inkscape, Krita, Clang itself, all using LLVM/Clang.
- If that's not possible/desirable (I can see that would be easier keep using Visual Studio for all Windows related builds).Is there a list of dependencies (including transitive dependencies!) that Stellarium uses somewhere where I can see if it's possible to build them with Visual Studio? 
- On the Visual Studio topic: VS2019 does support compiling for ARM64, but VS2022 is much preferrable because its codegen is much better for ARM64. Is it possible to switch to VS2022 at least for the ARM64 build?
- I saw some Github Actions workflows for building the Linux and MacOS builds, are there plans to move the Windows builds to Github Actions?

Some caveats: 
- MSYS2 does not support cross compiling for Windows on ARM64, which means that either the AppVeyor or Gitub Actions runners will have to run on native ARM64 Windows. The most obvious choice is to ask Microsoft for one of these: https://learn.microsoft.com/en-us/windows/arm/dev-kit/#device-specifics
- Visual Studio 2022 does support cross compiling for Windows on ARM64 from x64 Windows, including Qt 5 and 6 (I don't know the status of this for QtWebEngine), but that means that all the dependencies pre-built for Windows need to also be cross compiled for ARM64 with Visual Studio which is a tall order. It's a big list of dependencies. 

Most of the dependencies are already being built for ARM64 using MSYS2, that's why I suggested MSYS2 in the first place. But I understand that the whole plugin ecosystem is already being developed with Visual Studio.

**Reply by alex-w** [reply]
See preinstalled software - https://www.appveyor.com/docs/windows-images-software/

I guess reading AppVeyor docs can give answer for most requested questions. 

> I saw some Github Actions workflows for building the Linux and MacOS builds, are there plans to move the Windows builds to Github Actions?

No, because we use AppVeyor environment for signing software + Github Actions has bad support the dependencies for Stellarium, e.g. Qt framework (of course we can add more actions to install Qt, etc.).

**Reply by gzotti** [reply]
I think not using MSYS2 is just a result of having found one working solution for this platform (QtCreator/MSVC2012? and then updated as Qt required), and else caring for content. The "plugin ecosystem" is IMO rather a build-time decision, as dynamic plugins have last worked with Qt4 (correct me if I am wrong!), so at least I have never used dynamic plugins. If you manage to reactivate MSYS2, allow VS2022 also on W11/x64 without breaking the other options (Qt5.12+ (32/64, Win7+), Qt6; VS2017, 2019), we will likely not object. 

EDIT: Alexander has deeper insights into the build environments, so his answers are more "official".  :-)

**Comment by alex-w** [comment]
> I think not using MSYS2 is just a result of having found one working solution for this platform (QtCreator/MSVC2012? and then updated as Qt required), and else caring for content. The "plugin ecosystem" is IMO rather a build-time decision, as dynamic plugins have last worked with Qt4 (correct me if I am wrong!), so at least I have never used dynamic plugins. If you manage to reactivate MSYS2, allow VS2022 also on W11/x64 without breaking the other options (Qt5.12+ (32/64, Win7+), Qt6; VS2017, 2019), we will likely not object.

At some stage, I just skipped the building environment for testing builds for MinGW/MSYS2 and in some time later we got partially broken support this platform. According to patches for MSYS2 support, the difference with MSVC environment is not very critical, and we may restore support the MSYS2/MinGW environments.

I hope @hmartinez82 will prepare PR for it.

**Reply by hmartinez82** [reply]
I can work on getting the PR for MSYS2, the issue is the project getting one of these (https://www.appveyor.com/docs/windows-images-software), or a Windows on ARM64 VM and installing the AppVeyor runner so we can get the Windows on ARM64 builds going. Github Actions nor AppVeyor provide Windows on ARM runners

**Comment by hmartinez82** [comment]
Some good news. It turns out that we can just build Stellarium for Windows on ARM with Visual Studio itself. AppVeyor has Qt 6.5.3 for Windows ARM prebuilt! This PR adds the building for Windows on ARM https://github.com/Stellarium/stellarium/pull/3447.
Next steps:
- Get Spout2 compiling for Windows on ARM. Work is already under way: https://github.com/leadedge/Spout2/pull/104
- Add changes to support creating Stellarium Windows on ARM installers

**Comment by alex-w** [comment]
Please check our arm64 package for Windows - https://github.com/Stellarium/stellarium-data/releases/tag/weekly-snapshot
